### ISSUE TITLE
Nested*DepthRule

### ISSUE THREAD

**Issue by vanniktech** [detekt/detekt#248]
It would be cool to have Nested*DepthRules that check the depth of constructs such as `when`, `if`, `for`, `try` and allow them to be only that deep.

Given `NestedIfDepth` and a max depth of 1 (the default should probably be 2) the following code would then get flagged:

```kotlin
val x = 5
val y = 4
if (x == 5) {
  if (y == 5) { // This line will get flagged
    // Something.
  }
}
```

while 

```kotlin
val x = 5
val y = 4
if (x == 5) {
  // Something.
}
```

would be valid.

[Checkstyle reference NestedIfDepth](http://checkstyle.sourceforge.net/config_coding.html#NestedIfDepth) - [Checkstyle reference NestedForDepth](http://checkstyle.sourceforge.net/config_coding.html#NestedForDepth) - [Checkstyle reference NestedTryDepth](http://checkstyle.sourceforge.net/config_coding.html#NestedTryDepth)

**Comment by arturbosch** [comment]
Hm I would say this is a more custom rule and that the default NestedBlockDepth of three is enought. Maybe a second detekt supported rule set like `controversial` could be a good fit for this?

**Comment by vanniktech** [comment]
I'm always a fan of having lots of good rules to choose from. Also if we choose the default values carefully you shouldn't get the many reports back.

**Comment by Mauin** [comment]
Agreed, there is no such thing as having too many rules. Someone or some team will find each rule useful. We should just be sensible about turning on rules in the default ruleset to not overwhelm users.

**Comment by bobisageek** [comment]
Anyone mind if I take a shot at these rules? I've got three of them implemented (they seem to be working from the specs, but I haven't really tested them on the CLI yet).

If that's cool, I'll probably have a few questions I'd want to get right before submitting the PR:
- Should they all go under complexity?
- Am I missing any rule implementation steps from the list: add the rule and the spec, add 'active' and any config params to the default config, add the rule to the appropriate provider, and add the documentation to the appropriate category's markdown file?
- Based on the convo, I assume we'd want these all off by default, but there's also the question of default allowable nesting levels - I'm thinking maybe 2 for `if` and `for`, 1 for `try` (feels like a `try` adds more complexity because of the possible implications to control flow), and I'm on the fence between 1 and 2 regarding `when`

**Comment by arturbosch** [comment]
Hey, sure go ahead, a PR would be awesome!
1. Complexity sounds good to me
2. our contributing guide should mention most things. You do not need to manually touch the default config and the markdown file. They are generated based in the KDoc of the rules. Just take a look how NestedBlockDepth is implemented.
3. Thresholds sound okay for now. I think having a PR will encourage the discussion on the thresholds :)

**Comment by 3flex** [comment]
@bobisageek how did you go with the new rules? It would be great to get these into detekt!
